webapp/packages/core-settings-user/src/UserSettingsService.ts

```diff
@@ -0,0 +1,134 @@
+/*
+ * CloudBeaver - Cloud Database Manager
+ * Copyright (C) 2020-2024 DBeaver Corp and others
+ *
+ * Licensed under the Apache License, Version 2.0.
+ * you may not use this file except in compliance with the License.
+ */
+import { action, makeObservable, observable, runInAction } from 'mobx';
+
+import { UserInfoResource } from '@cloudbeaver/core-authentication';
+import { injectable } from '@cloudbeaver/core-di';
+import type { ISettingsSource } from '@cloudbeaver/core-settings';
+import { StorageService } from '@cloudbeaver/core-storage';
+import { isNotNullDefined, parseJSONFlat } from '@cloudbeaver/core-utils';
+
+@injectable()
+export class UserSettingsService implements ISettingsSource {
+  private readonly settings: Map<string, any>;
+  private readonly changes: Map<string, any>;
+  private lastConfig: any;
+  private localSettings: Map<string, any>;
+
+  constructor(
+    private readonly userInfoResource: UserInfoResource,
+    private readonly storageService: StorageService,
+  ) {
+    this.lastConfig = null;
+    this.settings = new Map();
+    this.changes = new Map();
+    this.localSettings = new Map();
+
+    this.userInfoResource.onDataUpdate.addHandler(this.refreshConfig.bind(this));
+    this.userInfoResource.onUserChange.addHandler(this.refreshConfig.bind(this));
+
+    makeObservable<this, 'settings' | 'changes' | 'refreshConfig' | 'localSettings'>(this, {
+      refreshConfig: action,
+      changes: observable.shallow,
+      settings: observable,
+      localSettings: observable,
+    });
+
+    this.storageService.registerSettings(
+      'user_settings',
+      this.localSettings,
+      () => new Map(),
+      undefined,
+      () => this.refreshConfig(),
+    );
+  }
+
+  has(key: any): boolean {
+    return this.getSource().has(key) || this.changes.has(key);
+  }
+
+  isEdited(key?: any): boolean {
+    if (isNotNullDefined(key)) {
+      return this.changes.has(key);
+    }
+
+    return this.changes.size > 0;
+  }
+
+  isReadOnly(key: any): boolean {
+    return false;
+  }
+
+  getEditedValue(key: any): any {
+    if (this.changes.has(key)) {
+      return this.changes.get(key);
+    }
+
+    return this.getValue(key);
```

Comment:
```suggestion
    return this.changes.get(key) ?? this.getValue(key);
```

Reply:
`get` can return `undefined` or `null` theoretically

Reply:
here in settings we want to extract value only if source "has" it. In case value is null or undefined we want to return this value as actual value